### PR TITLE
Add parser and downloader factory for the upgrade case

### PR DESCRIPTION
**Pull request by drevell**
When we're doing template upgrades, we'll have a canonical location and a location type, and we'll need a `Downloader` to get the new template contents. This code does just that: `f(location_type, location) -> Downloader`

### CODE REVIEW

File: templates/common/templatesource/upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+		{
+			name:        "remote_git_https_no_subdir",
+			location:    "github.com/abcxyz/abc",
+			locType:     "remote_git",
+			gitProtocol: "https",
+			wantDownloader: &remoteGitDownloader{
+				canonicalSource: "github.com/abcxyz/abc",
+				cloner:          &realCloner{},
+				remote:          "https://github.com/abcxyz/abc.git",
+				subdir:          "",
```
Comment (capri-xiyue) [inline templates/common/templatesource/upgrade_test.go:52]: Nits: Do we need to explicitly set subdir as "". Same below

Comment (drevell) [reply templates/common/templatesource/upgrade_test.go]: Fixed.

File: templates/common/templatesource/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			`$`) // Anchor the end, must match the entire input
+)
+
+type upgradeDownloaderFactory func(_ context.Context, canonicalLocation, gitProtocol, destDir string) (Downloader, error)
+
+// ForUpgrade takes a location type and canonical location from a manifest file,
+// and returns a downloader that will download the latest version of that
+// template.
+func ForUpgrade(ctx context.Context, canonicalLocation, locType, gitProtocol, destDir string) (Downloader, error) {
+	factory, ok := upgradeDownloaderFactories[locType]
```
Comment (yolocs) [inline templates/common/templatesource/upgrade.go:55]: Definitely fancier than a `swtich` :)

**Review by yolocs — APPROVED** [review]

File: templates/common/templatesource/upgrade.go (hunk 2 above)
Comment (drevell) [reply templates/common/templatesource/upgrade.go:55]: 🎩

**Pull request merged**